ENH: drafted roadmap section of manuscript

tylerjereddy opened the pull request:
I tried to draft a roadmap section for the manuscript, which is obviously based pretty strongly on our actual roadmap. It may suffer somewhat from the list-like structure of the original as far as being nicely-flowing prose goes.

jarrodmillman commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +185,85 @@ \section*{Project organisation and community}
+for improvement for the project, as well as specific limitations and
+matters that require assistance moving forward.
+
+In general, we encourage changes that improvie clarity in the API of the library,
```
s/improvie/improve

jarrodmillman commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +185,85 @@ \section*{Project organisation and community}
+too much community code duplication with \texttt{numpy.fft}, which
+requires clarification of best choices and long-term plans.
+
+\texttt{scipy.integrate} now has a new ODE solver interface, and we should aim
```
s/now//

jarrodmillman commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +185,85 @@ \section*{Project organisation and community}
+to be updated to match the capabilities in other routines, and wavelet-related
+routines are somewhat out of place.
+
+\texttt{scipy.sparse} is fairly mature now, but there are many data
```
Maybe "While \texttt{scipy.sparse} is mature, there are ..."

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +185,85 @@ \section*{Project organisation and community}
+
+We're planning to remove \texttt{scipy.misc} as a public module, and
+rename \texttt{scipy.odr} to \texttt{scipy.regression} or
+\texttt{scipy.fitting}, and include \texttt{optimize.curve\_fit}. This 
```
Consider "relocate..." or "migrate... to this new home" or similar. "include" might suggest that `optimize.curve\_fit` is not presently included in SciPy.

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +185,85 @@ \section*{Project organisation and community}
+
+\texttt{scipy.ndimage} should be completely migrated to a data point
+model (values on a grid), instead of a pixel model (elements with centers),
+both of which have been used in the past, because the former is now
```
"the former" may be ambiguous here. Clearly you mean the first one you mentioned ("values on a grid", not necessarily the "former" choice of models temporally), but perhaps there is a better word choice given that you have just mentioned "the past".

tylerjereddy commented:
Rebased / force pushed and revised.

Things I'm not sure about:

1. SciPy tutorial page(s) traffic -- maybe a senior team member has access to that info if they also deem it pertinent here. For now, I just indicated that there are 15 sections to our tutorial along with a citation to the online resource.
2. For the typed memoryview clarification requested I added a short code listing / snippet -- as I recall there was some informal objection to code listings in the ms. during the planning stages so this may require some justification or just switching to an inline prose explanation somehow.

endolith commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+
+\begin{lstlisting}[language=Python]
+# example of a function that takes a single
+# 3 dimesional typed memoryview as an argument
```
"dimensional" typo

rgommers commented:
>  It may suffer somewhat from the list-like structure of the original as far as being nicely-flowing prose goes.

I think it would be good to limit the discussion to key future changes, and remove the rest. E.g. `fftpack`, `io`, `KDTree` are all not very interesting to a reader. Of the ones mentioned, these are the interesting entries I think: `interpolate`, `linalg`, `optimize`, `ndimage`, `signal`.

`stats` is missing. The addition of *sparse arrays* is also worth mentioning.

ev-br commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+
+\texttt{scipy.interpolate} could benefit from a set of new features including
+spline fitting routines with better user control; integration, differentiation,
+and arithmetic routines for splines; transparent tensor-product splines,
```
Not sure what integration and differentiation means here: this has been implemented. (The rest I totally agree with.)

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+
+In general, we encourage changes that improve clarity in the API of the library,
+but strongly discourage breaking backwards compatibility, given our position near
+the base of the scientific Python computing stack.
```
Good paragraph, but let's move it to the API/ABI evolution section

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+respectable, with overall project test coverage holding steady at
+around 76\% for the last six months (Figure~\ref{fig:coverage}),
+but some of the historical components of the code base
+would still benefit from increased test coverage. We are still striving
```
This can go to the testing/benchmarking section

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+around 76\% for the last six months (Figure~\ref{fig:coverage}),
+but some of the historical components of the code base
+would still benefit from increased test coverage. We are still striving
+to increase the number of high-quality SciPy usage tutorials beyond
```
probably should remove "high-quality", because that's not really what they are I think - they could use some work

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+	 automatically in our continuous integration suite using the
+	 \texttt{pytest-cov} and \texttt{gcov} libraries.}
+\label{fig:coverage}
+\end{figure}
```
nice figure, let's keep that in the testing section as well

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+
+\texttt{scipy.integrate} has a new ODE solver interface, and we should aim
+to gracefully deprecate the older API. Support for integrating complex-valued
+functions and integrating multiple intervals could be beneficial.
```
I'd remove the bit about the old API for sure. The rest is probably also not major enough to mention

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+\texttt{fftpack} has some single-precision floating point errors, a caching bug,
+and would benefit from the addition of the Bluestein algorithm. There's also
+too much community code duplication with \texttt{numpy.fft}, which
+requires clarification of best choices and long-term plans.
```
I propose to remove the first bit, and mention the overlap with numpy for both `fftpack` and `linalg` together

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+
+\texttt{scipy.io} would benefit from PCM float support, but we otherwise
+direct users with more sophisticated needs to domain-specific audio
+libraries.
```
let's remove this, too minor

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+\texttt{get\_lapack\_funcs} should always use flapack, and
+we should continue to wrap more LAPACK functions. Ideally, we'd also
+add type-generic wrappers in the Cython BLAS and LAPACK interfaces, and
+transform many of the linear algebra routines into gufuncs.
```
let's remove the `flapack` stuff, not very user relevant. The type-generic wrappers thing we could keep here, but only if we rephrase it so that it makes sense to the average reader

rgommers replied:
The most important thing for `linalg` I think is to start upgrading to more recent LAPACK versions.

ilayn replied:
I think for the average reader it is more important to mention that they can access the LAPACK funcs via either `get_lapack_funcs` or via `sp.linalg.lapack.<whatever>`.

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+\texttt{scipy.fitting}, and relocate \texttt{optimize.curve\_fit}. This 
+module will then provide a home for other fitting functionality.
+\texttt{scipy.optimize} would benefit from a few more good global optimizers 
+as well as large-scale optimizers.
```
I propose just to keep the `optimize` part. E.g.: The `optimize` module will keep evolving, we expect to add several new global optimization methods with improved performance over the existing ones, as well as add large-scale optimizers.

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+model (values on a grid), instead of a pixel model (elements with centers),
+both of which have been used in the past, because the data point model is now
+proving more effective in practice. We also need to standardize the
+\texttt{scipy.ndimage.morphology} API.
```
let's keep the pixel-vs-point here, and remove the morphology bit. @stefanv has a draft section on pixel-vs-point, we can align these two sections later after merging

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+to be reduced; B-splines functionality should be moved to \texttt{interpolate},
+\texttt{firwin2} can be merged into \texttt{firwin}; Second Order Sections need
+to be updated to match the capabilities in other routines, and wavelet-related
+routines are somewhat out of place.
```
I suggest to mention `signal` in the section that's now about `interpolate` (talk about all spline topics there). SOS is a good topic perhaps. I would remove the rest.

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+formats and the community continues to move forward here so we will aim
+to take advantage of their improvements while preserving backwards
+compatibility to the extent that it is possible. \texttt{sparse.linalg}
+may benefit from wrappers for PROPACK for faster sparse SVD computation.
```
Two things I'd mention here: sparse arrays (https://github.com/pydata/sparse) and (perhaps) PROPACK.

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+may benefit from wrappers for PROPACK for faster sparse SVD computation.
+
+For \texttt{scipy.spatial}, KDTree will be removed, and cKDTree will be 
+renamed to KDTree in a backwards-compatible way. \texttt{distance\_wrap.c} 
```
I wouldn't mention the `KDTree` removal, users won't notice since it becomes an alias of `cKDTree`. There's nothing in `spatial` important enough to mention AFAIK.

rgommers replied:
Sorry changed my mind: for `spatial` we can mention `spatial.transforms` with support for rotation matrices.

rgommers commented on scipy-1.0/paper.tex:
```diff
@@ -185,6 +186,114 @@ \section*{Project organisation and community}
+renamed to KDTree in a backwards-compatible way. \texttt{distance\_wrap.c} 
+likely needs a rewrite in Cython. \texttt{scipy.special} needs precision
+improvements for hypergeometric functions, parabolic cylinder functions,
+and spheroidal wave functions.
```
we may keep the `special` stuff, I think it's a big topic.

ilayn commented:
You might want to add some more options for the code listings to render them a bit prettier, this is the current rendering 

<hr>

![image](https://user-images.githubusercontent.com/1303842/43159189-f02f995a-8f81-11e8-8aec-89d5e5fc3dca.png)
<hr>
If you also add `textcomp` package and add the following 

```latex
\usepackage{listings, textcomp}
    
\lstset{ %
  basicstyle=\ttfamily\footnotesize,  % size of fonts used for the code
  breaklines=true,   % automatic line breaking only at whitespace
  captionpos=b,   % sets the caption-position to bottom
  commentstyle=\color{gray},  % comment style
  keywordstyle=\color{blue},  % keyword style
  stringstyle=\color{red},  % string literal style
  upquote=true  %straight single quotes (requires textcomp)
}
```
it becomes 

![image](https://user-images.githubusercontent.com/1303842/43159236-1a903eac-8f82-11e8-9b38-e2aa9e7d7437.png)


I am not suggesting that this is the optimal output but in my opinion it is a bit better in terms of blending into text. Maybe a frame or enabling line numbers can make it more obvious .

tylerjereddy commented:
I tried to address most of the revisions, but noticed I forgot the typo pointed out by @endolith and the recent code format thing from @ilayn.

The revisions were pretty major & rebasing a little messy so good to double check.

tylerjereddy commented:
Ok, I think I've addressed those two points now as well & added @ilayn as co-author on latest commit since I copy-pasted the suggested  adjustment to code listing appearance.

rgommers commented:
LGTM now, merged. Thanks Tyler

ilayn commented:
Just for the record, you can skip the coauthor step :) My contribution is really minor stylistic stuff
